NativeAOT-LLVM: merge PR 67238 - JIT: Refactor call argument representation

yowl opened the pull request:
This PR merges runtime PR https://github.com/dotnet/runtime/pull/67238 where the call args are refactored into `callNode->gtArgs`  .  Most of the work is in `lowerCallToShadowStack`.

The "special" LLVM args, shadow stack, and return slot are now inserted in reverse order at the end using `PushFront`.

jakobbotsch committed:
JIT: Refactor call argument representation (#67238)

This refactors the JIT's representation of call arguments. It replaces
`GenTreeCall::Use` and `fgArgTabEntry` with a single class `CallArg`.
`CallArg` always contains space for ABI information and contains two
intrusive linked list nodes: one for all args (similar to current
`gtCallArgs`) where all standard arguments are always in argument order,
and one for late args (similar to current `gtCallLateArgs`) that may be
reordered. The late args list may also not contain all arguments.

`fgArgInfo` is also replaced by a new class `CallArgs` that is stored
inline in `GenTreeCall`. It encapsulates all handling of arguments
(insertion/removal). The change also begins treating the 'this' argument
as a normal argument with `CallArgs` providing convenient access to it
when necessary.

The main benefit of this change is to avoid keeping track of the side
table `fgArgInfo` and having to scan through this side table repeatedly
when we need to query argument information. In addition it gives more
convenient ways to access well known arguments like 'this', the ret
buffer, VSD cell, etc. Finally, it also serves as a nice clean-up in the JIT.

yowl committed:
Merge commit 'dbf5c58a2e90a5dbe52248c5a4cad5e9aa6a4741' into pr67238

# Conflicts:
#	src/coreclr/jit/compiler.h
#	src/coreclr/jit/gentree.cpp
#	src/coreclr/jit/importer.cpp
#	src/coreclr/jit/lower.cpp
#	src/coreclr/jit/morph.cpp

yowl committed:
trivially resolve conflicts

yowl committed:
put back closing endif

yowl commented:
@SingleAccretion `GT_PUTARG_TYPE` is not removed in this PR, but it is the intention that it is removed later.  Are we going to remove `GT_PUTARG_TYPE` when lowering calls, i.e. remove calls to `gtNewPutArgType`, with this PR or later when `GT_PUTARG_TYPE` is removed upstream?

SingleAccretion commented:
To remove `GT_PUTARG_TYPE` we need https://github.com/dotnet/runtime/pull/68748 first, otherwise there is no place to store the signature types, so I suppose the latter.

yowl committed:
Clear errors in gentree.cpp:

- remove customregister cases for Wasm for R2RIndirectionCell and ValidateIndirectCallTarget - might need revisiting as not sure we will hit these.
- temporary function for arg message for gtDispLIRNode with GT_PUTARG_TYPE

yowl committed:
remove non llvm file compilation errors

- Remove compiler.codegen field
- Change assert for HasRetBufArg - to iciCall->gtArgs.HasRetBuffer() - might have to revisit this

yowl commented:
As it looks harder than just `callNode->gtCallArgs = nullptr` to start with new args, I'm thinking there are a couple of options for `lowerCallToShadowStack` 

* capture the existing set of args somehow, remove the existing args using `callNode->gtArgs.Remove()` and add new ones
* create a new `GenTreeCall` node and build using the old one as the source (instead of `argInfo->ArgTable`)

Did you have any thoughts about what would be best?

SingleAccretion commented:
Recreating the argument list should no longer be necessary. Only the underlying nodes need to be changed.

yowl committed:
remove GenTreeCall

yowl committed:
define MIN_ARG_AREA_FOR_CALL for LLVM/Wasm

yowl commented:
I have a question about morphing the args.  I'm hitting the assert at 

https://github.com/yowl/runtimelab/blob/d1a050268b7a41d4e529baf28203576825a8200b/src/coreclr/jit/morph.cpp#L8910

Because `this` had not been morphed into a temp.  Tracing the execution, this seems to be because, although `this` is identified as needing a temp and get `m_needTmp` set `true`, the execution goes into 

https://github.com/yowl/runtimelab/blob/pr67238/src/coreclr/jit/morph.cpp#L1589-L1590

And skips the creation of the temp.  The comment above this says " On x86 and other archs that use push instructions to pass arguments:" .

We set `REG_STK` for `this` at 

https://github.com/yowl/runtimelab/blob/pr67238/src/coreclr/jit/morph.cpp#L3020

I don't know if we should be looking to see how x86 gets past the assert, or for `TARGET_WASM` we should be ignoring the `arg.AbiInfo.GetRegNum() == REG_STK` check and creating the temp?

(dump attached)
[dump.txt](https://github.com/dotnet/runtimelab/files/10472492/dump.txt)

SingleAccretion commented:
Upstream assumes all non-local `this` args are forced into temps because they are always late, which is in turn assumed because all targets pass `this` in a register.

We used to get away with this because of the special handling of `this` that forced it into a register, but now that special handling is gone.

Now, let's consider the case where `this` would be passed on stack via `push`. We would still not be able to evaluate it into a temp, since that would change `push` order:
```
push this
push <other args>
->
push <other args>
push this
```
Essentially, to put `push`ed args into temps, `EvalArgsToTemps` would need to do what `fgMakeOutgoingStructArgCopy` does i. e. put the expression into a `COMMA` instead of using the late arg list. It does not do that.

Let's now consider the shadow stack ABI. Here we don't actually do `push`, we just store the arguments into the outgoing area right before the call, so it is not a problem if we "reorder" the arguments like above.

So I think we can just put ` && !defined(TARGET_WASM)` to the `#if !FEATURE_FIXED_OUT_ARGS` in `EvalArgsToTemps`.

yowl commented:
Thanks, that clears that problem.  I'm going back to something earlier:

> Only the underlying nodes need to be changed.

I'm having a spot of bother with this because, for arguments on the shadow stack, I'm attempting to remove them from `gtArgs` with `Remove`.  Firstly that doesn't like being called when `m_abiInformationDetermined == true`, but if I ignore that for now, it only removes the early arg, and when a temp has been created, the `CallArg` has both an early and late node by virtue of the code in `EvalArgsToTemps` :
```
  if (setupArg != nullptr)
  {
      noway_assert(arg.GetEarlyNode() == argx);
      arg.SetEarlyNode(setupArg);
  }

  arg.SetLateNode(defArg);
  *lateTail = &arg;
  lateTail  = &arg.LateNextRef();
```
Am I going to need to adjust `Remove` for us to remove from the late arg list as well, or am I on the wrong path?

SingleAccretion commented:
It seems like the right direction.

We'd need a different helper though, as we cannot really use `Remove` as-is because it would mean the algorithm is going to be quadratic. The helper would be something like `RemoveAfter(CallArg* arg)` (basically a wrapper over `SetNext`).

Removing late args on the fly seems a bit hard since they're unordered (the list could be: `arg3 -> arg1 -> arg2`, etc), but we can just null out the entire list by transferring late nodes to early ones, doing `SetLateNext(nullptr)` and nulling out `m_lateHead`.

yowl commented:
I was hoping I could get rid of the siginfo determination of call signature and rely on the args, but that's not going to be the case it seems (shorts->ints is one problem).  We will keep the siginfo code, that's as you expect?

SingleAccretion commented:
We need the signature for helpers since they aren't precisely tracked. For user/indirect calls, I'd expect we could use the fields added in https://github.com/dotnet/runtime/pull/68748 to track them, with a little change to use `CorInfoType` instead of `var_types` because we need to distinguish pointers from non-pointers.

yowl committed:
lower calls using gtArgs

yowl committed:
whitespace cleanup

yowl commented:
cc @dotnet/nativeaot-llvm

SingleAccretion commented on src/coreclr/jit/compiler.h:
```diff
@@ -8567,10 +7742,9 @@ class Compiler
-    CodeGenInterface* codeGen;
-
 #ifndef TARGET_WASM
     // Record the instr offset mapping to the generated code
+    CodeGenInterface* codeGen;
 
     jitstd::list<IPmappingDsc> genIPmappings;
```
```suggestion
    CodeGenInterface* codeGen;

    // Record the instr offset mapping to the generated code
    jitstd::list<IPmappingDsc> genIPmappings;
```

SingleAccretion commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -1840,6 +1842,39 @@ void CallArgs::Remove(CallArg* arg)
 }
 
+#if TARGET_WASM
+void CallArgs::RemoveAfter(CallArg* arg)
```
This needs a [standard function header](https://github.com/dotnet/runtimelab/blob/24524863c43f97a42a55466b5de4cee3fd6d22fb/src/coreclr/jit/conventions.txt#L57-L81).

Also, `RemovedWellKnownArg` needs to be called for any removed args.

yowl replied:
Thanks, added both

SingleAccretion commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -1840,6 +1842,39 @@ void CallArgs::Remove(CallArg* arg)
+    
+    m_lateHead = nullptr;
+}
+
```
```suggestion
```

SingleAccretion commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -1840,6 +1842,39 @@ void CallArgs::Remove(CallArg* arg)
+    arg->SetNext(arg->GetNext()->GetNext());
+}
+
+void CallArgs::MoveLateToEarly()
```
Ditto: comment header.

yowl replied:
Do we need something in these comments to say it applies to LLVM lowering?

SingleAccretion replied:
No need I think. The `#ifdef` is self-explanatory and the functions are general-purpose.

SingleAccretion commented on src/coreclr/jit/llvmlower.cpp:
```diff
@@ -597,13 +597,10 @@ void Llvm::lowerRethrow(GenTreeCall* callNode)
-    callNode->gtCallArgs = _compiler->gtNewCallArgs(excObjAddr);
-
-    _compiler->fgInitArgInfo(callNode);
+    callNode->gtArgs.InsertAfterThisOrFirst(_compiler, excObjAddr);
```
Why not `PushFront`?

SingleAccretion commented on src/coreclr/jit/llvmlower.cpp:
```diff
@@ -785,48 +782,15 @@ void Llvm::lowerReturn(GenTreeUnOp* retNode)
+// LVVM Arg layout:
+//    - Shadow stack (if required)
+//    - Return slot (if required)
+//    - Generic context (if required)
+//    - args passed as LLVM parameters (not on the shadow stack)
```
```suggestion
//    - Generic context (if required)
//    - Args passed as LLVM parameters (not on the shadow stack)
//
```

SingleAccretion commented on src/coreclr/jit/llvmlower.cpp:
```diff
@@ -775,7 +772,7 @@ void Llvm::lowerReturn(GenTreeUnOp* retNode)
 // This method has two primary objectives:
-//  1) Transfer the information about the arguments from arg info to explicit
+//  1) Transfer the information about the arguments from gtArgs to explicit
 //     PutArgType nodes, to make it easy for codegen to consume it. Also, get
 //     rid of the late/non-late argument distinction, by sorting the inserted nodes
 //     in the original evaluation order, matching that of them in the signature.
```
```suggestion
//     PutArgType nodes, to make it easy for codegen to consume it. Also, all
//     of the late argument nodes are moved (back) to the early list.
```

SingleAccretion commented on src/coreclr/jit/llvmlower.cpp:
```diff
@@ -997,36 +974,31 @@ void Llvm::failUnsupportedCalls(GenTreeCall* callNode)
+    for (CallArg& callArg : callNode->gtArgs.Args())
     {
-        for (GenTree* operand : callNode->Operands())
+        GenTree* operand = callArg.GetNode();
```
I think the whole:
```
        GenTree* operand = callArg.GetNode();
        if (operand->IsArgPlaceHolderNode() || !operand->IsValue())
        {
            // Either of these situations may happen with calls.
            continue;
        }
        if (operand == callNode->gtControlExpr || operand == callNode->gtCallAddr)
        {
            // vtable target or indirect target
            continue;
        }
```
Can just be deleted as it's a no-op. We only care about blocking VSD.

yowl replied:
Indeed.

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -3141,6 +3095,12 @@ void CallArgs::AddFinalArgsAndDetermineABIInfo(Compiler* comp, GenTreeCall* call
 }
 
+
+// normally in target<arch>.h but llvm/wasm does not have that file and seems overkill for this one constant
```
Hmm, but we do have `targetwasm.h`?

yowl replied:
/shuffles feet/

yowl replied:
moved to `targewasm.h`

SingleAccretion commented on src/coreclr/jit/llvmlower.cpp:
```diff
@@ -929,40 +895,51 @@ void Llvm::lowerCallToShadowStack(GenTreeCall* callNode)
+
+    callNode->gtArgs.MoveLateToEarly();
+
+    // Insert special args after finished iterating >gtArgs.Args
```
To be honest, I liked the setup where we were inserting these in-order at the top more. But it's not a big deal.

yowl replied:
I just assumed it was going to be a problem with iterator.  I could do the iteration manually, starting from the last of shadow stack or return slot.  Do you want to see what that looks like?

SingleAccretion replied:
>  Do you want to see what that looks like?

It would be appreciated.

I don't think manual `GetNext`-style iteration is a big downside, since we already effectively depend on the argument list being linked with `RemoveAfter`.

yowl replied:
Last commit has this.  There's some long compound ways to get the head of the list `callNode->gtArgs.Args().begin().GetArg()`.  I could add a `GetHead` method to make that look nicer?

SingleAccretion replied:
Not worth it at present I think, the use in return lowering could be replaced with `GetArgByIndex(0)`, and then only only one is left (modulo upstream)

yowl replied:
>  I think, the use in return lowering could be replaced with GetArgByIndex(0)

I added a commit for this.  If it goes green I will ping Jan.  Thanks for the review!

SingleAccretion commented on src/coreclr/jit/llvmlower.cpp:
```diff
@@ -929,40 +895,51 @@ void Llvm::lowerCallToShadowStack(GenTreeCall* callNode)
         argIx++;
     }
+
+    callNode->gtArgs.MoveLateToEarly();
```
Can we do this before at the top, before rewriting the argument list? It'd be a bit clearer I think.

yowl replied:
Yes, done

yowl committed:
address feedback, move things around, add comments, delete dead code.

yowl committed:
Update src/coreclr/jit/llvmlower.cpp

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update src/coreclr/jit/compiler.h

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update src/coreclr/jit/gentree.cpp

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update src/coreclr/jit/llvmlower.cpp

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

SingleAccretion commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -1840,6 +1842,57 @@ void CallArgs::Remove(CallArg* arg)
+    {
+        assert(m_head != nullptr);
+
+        m_head = m_head->GetNext();
```
`RemovedWellKnownArg` is needed here too.

yowl committed:
RemovedWellKnownArg for head removal

yowl committed:
Merge branch 'pr67238' of https://github.com/yowl/runtimelab into pr67238

yowl committed:
add shadow stack and return slot in the order they appear

SingleAccretion reviewed (approved):
LGTM, thank you!

yowl committed:
simplify getting first arg.

yowl commented:
@jkotas This looks good to go, thanks!

jkotas commented:
Great work! Thank you!
